src/util/container.h

```diff
@@ -305,3 +316,157 @@ class LRUCache
 	// we can't use std::deque here, because its iterators get invalidated
 	std::list<K> m_queue;
 };
+
+/*
+	Map that can be safely modified (insertion, deletion) during iteration
+	Caveats:
+	- be careful of nullptr elements when using iter(), those are ones already deleted
+	- size() and empty() will not function during iteration
+	- not thread-safe in any way
+*/
+
+template<typename K, typename V>
+class ModifySafeMap
+{
+public:
+	static_assert(std::is_pointer<V>::value, "Data structure requires pointer values");
+	// TODO is_default_constructible + is_convertible to bool?
+
+	ModifySafeMap() = default;
+	~ModifySafeMap() { assert(!m_iterating); }
+
+	V get(const K &key) const {
+		if (m_iterating) {
+			auto it = m_new.find(key);
+			if (it != m_new.end())
+				return it->second;
+		}
+		auto it = m_values.find(key);
+		return it == m_values.end() ? nullptr : it->second;
+	}
+
+	void put(const K &key, V value) {
+		assert(value);
+		if (m_iterating) {
+			m_new.emplace(key, value);
+		} else
+			m_values.emplace(key, value);
+	}
+
+	bool remove(const K &key) {
+		bool ret = false;
+		if (m_iterating) {
+			auto it = m_new.find(key);
+			if (it != m_new.end()) {
+				m_new.erase(it);
+				ret = true;
+			}
+		}
+		auto it = m_values.find(key);
+		if (it == m_values.end())
+			return ret;
+		if (m_iterating) {
+			it->second = nullptr;
+			m_garbage++;
+		} else {
+			m_values.erase(it);
+		}
+		return true;
+	}
+
+	size_t size() const {
+		if (m_iterating) {
+			// This is by no means impossible to determine, it's just annoying
+			// to code and we happen to not need this.
+			return unknown;
+		}
+		assert(m_new.empty());
+		if (m_garbage == 0)
+			return m_values.size();
+		size_t n = 0;
+		for (auto it : m_values)
+			n += it.second == nullptr ? 0 : 1;
+		return n;
+	}
+
+	bool empty() const {
+		if (m_iterating)
+			return false; // maybe
+		if (m_garbage == 0)
+			return m_values.empty();
+		for (auto it : m_values) {
+			if (!(it.second == nullptr))
+				return false;
+		}
+		return true;
+	}
+
+	auto iter() { return IterationHelper(this); }
+
+	void clear() {
+		if (m_iterating) {
+			for (auto &it : m_values)
+				it.second = nullptr;
+			m_garbage = m_values.size();
+		} else {
+			m_values.clear();
+			assert(m_new.empty());
+			m_garbage = 0;
+		}
+	}
+
+protected:
+	void merge_new() {
+		assert(!m_iterating);
+		if (!m_new.empty()) {
+			m_new.merge(m_values); // entries in m_new take precedence
```

Comment:
Oh nice, C++17 merge.

Should probably call the rvalue overload:
```suggestion
			m_new.merge(std::move(m_values)); // entries in m_new take precedence
```

Reply:
merge modifies the source so I don't know what the difference would be